### PR TITLE
Feature/cubed sphere cell meshgen

### LINKED ISSUE #45: Cell-centroid mesh generator
Need to write a new mesh-generator based on cell-centroid cubed sphere grid.

### PR DESCRIPTION
**Pull request by odlomax**
This adds a single PE mesh generator for a cubed sphere grid with cell-centroid xy positions (as opposed to nodal xy positions).

Every node point added to the mesh has a unique global id. The global ids for "owned nodes" match those of an equivalent resolution nodal grid.

Code has been added to make sure that the remote indices of ghost nodes match those of owned nodes across tile-boundaries. This can been seen by inspecting the gmsh outputs. Example shown below.

The nodes created for this mesh should have the same xy positions as a nodal grid with the same resolution. The test checks for this.

Note: the previous mesh generator as temporarily been renamed FV3CubedSphereGenerator. The associated `.h` and `.cc` files, and the removal thereof, should be considered as a separate issue.

I have (perhaps unwisely) updated atlas to the head of develop. The files changed be *me* in this PR are:

src/atlas/CMakeLists.txt
src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc
src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.h
src/atlas/meshgenerator/detail/FV3CubedSphereMeshGenerator.h
src/atlas/meshgenerator/detail/FV3CubedSphereMeshGenerator.h
src/tests/grid/test_cubedsphere.cc

Please ignore the others!

Closes #45 

remote_idx

### CODE REVIEW

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 1]
```diff
@@ -42,298 +45,348 @@ namespace meshgenerator {
 
 // -------------------------------------------------------------------------------------------------
 
-CubedSphereMeshGenerator::CubedSphereMeshGenerator( const eckit::Parametrisation& p ) {}
+CubedSphereMeshGenerator::CubedSphereMeshGenerator(const eckit::Parametrisation& p) {}
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:48]: I think that the local coding style is to have a space between brackets and things within them.

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 2]
```diff
@@ -42,298 +45,348 @@ namespace meshgenerator {
...
-    // Tile 5
-    for ( idx_t ix = 1; ix < N; ix++ ) ghostToOwnedTij.push_back(TijPair{{4, ix, N}, {0, 0, N - ix}});
-    for ( idx_t iy = 0; iy < N; iy++ ) ghostToOwnedTij.push_back(TijPair{{4, N, iy}, {5, 0, iy}});
-
-    // Tile 6
-    for ( idx_t ix = 0; ix < N; ix++ ) ghostToOwnedTij.push_back(TijPair{{5, ix, N}, {0, ix, 0}});
-    for ( idx_t iy = 1; iy < N; iy++ ) ghostToOwnedTij.push_back(TijPair{{5, N, iy}, {1, N - iy, 0}});
-
-
+void CubedSphereMeshGenerator::generate(const Grid& grid, Mesh& mesh) const {
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:56]: ditto

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc [hunk 3]
```diff
@@ -42,298 +45,348 @@ namespace meshgenerator {
...
-    auto flags         = array::make_view<int, 1>( nodes.flags() );
+  if (gridStagger != "C") {
+    throw_Exception("CubedSphereMeshGenerator can only work with a"
+      "cell-centroid grid. Try FV3CubedSphereMeshGenerator instead.");
+  }
 
-    // Loop over entire grid
-    // ---------------------
+  // Partitioner
+  const auto partitioner = grid::Partitioner("checkerboard", 1);
```
Comment (MarekWlasak) [inline src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:76]: I presume the above line is going to change.

**Review by MarekWlasak — APPROVED** [review]

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc (hunk 1 above)
Comment (odlomax) [reply src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:48]: I agree. That said, I'd have to redo almost every other line in my code. I think Willem has a script that correctly formats this.

File: src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc (hunk 3 above)
Comment (odlomax) [reply src/atlas/meshgenerator/detail/CubedSphereMeshGenerator.cc:76]: Yep, very much an hangover from the last version of this!

**Pull request merged**